source/Octopus.Tentacle.Tests.Integration/ClientFileTransferRpcRetryTimeoutsWork.cs

```diff
@@ -0,0 +1,157 @@
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading.Tasks;
+using FluentAssertions;
+using Halibut;
+using NUnit.Framework;
+using Octopus.Tentacle.Tests.Integration.Support;
+using Octopus.Tentacle.Tests.Integration.Util;
+using Octopus.Tentacle.Tests.Integration.Util.Builders;
+using Octopus.Tentacle.Tests.Integration.Util.Builders.Decorators;
+using Octopus.Tentacle.Tests.Integration.Util.TcpTentacleHelpers;
+using Octopus.Tentacle.Tests.Integration.Util.TcpUtils;
+
+namespace Octopus.Tentacle.Tests.Integration
+{
+    /// <summary>
+    /// These tests make sure that we can cancel or walk away (if code does not cooperate with cancellation tokens)
+    /// from RPC calls when they are being retried and the rpc timeout period elapses.
+    /// </summary>
+    [IntegrationTestTimeout]
+    public class ClientFileTransferRpcRetryTimeoutsWork : IntegrationTest
+    {
+        [Test]
+        [TestCase(TentacleType.Polling, false)] // Timeout an in-flight request
+        [TestCase(TentacleType.Polling, true)] // Timeout trying to connect
+        [TestCase(TentacleType.Listening, false)] // Timeout an in-flight request
+        [TestCase(TentacleType.Listening, true)] // Timeout trying to connect
+        public async Task WhenRpcRetriesTimeOut_DuringUploadFile_TheRpcCallIsCancelled(TentacleType tentacleType, bool stopPortForwarderAfterFirstCall)
+        {
+            PortForwarder portForwarder = null!;
+            using var clientAndTentacle = await new ClientAndTentacleBuilder(tentacleType)
+                // Set a short retry duration so we cancel fairly quickly
+                .WithRetryDuration(TimeSpan.FromSeconds(15))
+                .WithPortForwarderDataLogging()
+                .WithResponseMessageTcpKiller(out var responseMessageTcpKiller)
+                .WithTentacleServiceDecorator(new TentacleServiceDecoratorBuilder()
+                    .LogCallsToFileTransferService()
+                    .CountCallsToFileTransferService(out var fileTransferServiceCallCounts)
+                    .RecordExceptionThrownInFileTransferService(out var fileTransferServiceException)
+                    .DecorateFileTransferServiceWith(d => d
+                        .BeforeUploadFile((service, _, _) =>
+                        {
+                            service.EnsureTentacleIsConnectedToServer(Logger);
```

Comment:
@nathanwoctopusdeploy Should this be within the `if == null` just below? Otherwise I think this will be trying to ensure the tentacle is connected possibly when the port-forwareder is down.